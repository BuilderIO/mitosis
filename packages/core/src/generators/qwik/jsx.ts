import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
