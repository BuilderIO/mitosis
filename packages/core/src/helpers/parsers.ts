import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
