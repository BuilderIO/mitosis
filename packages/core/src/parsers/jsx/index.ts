export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
