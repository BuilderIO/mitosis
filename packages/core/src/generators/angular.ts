import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { getProps } from '../helpers/get-props';
import { kebabCase } from 'lodash';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { indent } from '../helpers/indent';

export interface ToAngularOptions extends BaseTranspilerOptions {}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAngularOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToAngular(item, options))
      .join('\n')}</div>`;
  },
};

export const blockToAngular = (
  json: MitosisNode,
  options: ToAngularOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<ng-content></ng-content>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `<ng-container *ngFor="let ${
      json.properties._forName
    } of ${stripStateAndPropsRefs(json.bindings.each as string)}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else if (json.name === 'Show') {
    str += `<ng-container *ngIf="${stripStateAndPropsRefs(
      json.bindings.when as string,
    )}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else {
    str += `<${json.name} `;

    // TODO: spread support for angular
    // if (json.bindings._spread) {
    //   str += `v-bind="${stripStateAndPropsRefs(
    //     json.bindings._spread as string,
    //   )}"`;
    // }

    for (const key in json.properties) {
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === '_spread') {
        continue;
      }
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (
          event === 'change' &&
          json.name === 'input' /* todo: other tags */
        ) {
          event = 'input';
        }
        // TODO: proper babel transform to replace. Util for this
        const finalValue = removeSurroundingBlock(
          useValue.replace(/event\./g, '$event.'),
        );
        str += ` (${event})="${finalValue}" `;
      } else if (key === 'ref') {
        str += ` #${useValue} `;
      } else {
        str += ` [${key}]="${useValue}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToAngular(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

export const componentToAngular =
  (options: ToAngularOptions = {}): Transpiler =>
  ({ component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }

    const props = getProps(component);

    const refs = Array.from(getRefs(json));
    mapRefs(json, (refName) => `this.${refName}.nativeElement`);

    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    let template = json.children.map((item) => blockToAngular(item)).join('\n');
    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      valueMapper: (code) =>
        stripStateAndPropsRefs(code, { replaceWith: 'this.' }),
    });

    let str = dedent`
    import { Component ${refs.length ? ', ViewChild, ElementRef' : ''}${
      props.size ? ', Input' : ''
    } } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      selector: '${kebabCase(json.name || 'my-component')}',
      template: \`
        ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
      \`,
      ${
        css.length
          ? `styles: [
        \`${indent(css, 8)}\`
      ],`
          : ''
      }
    })
    export default class ${component.name} {
      ${Array.from(props)
        .map((item) => `@Input() ${item}: any`)
        .join('\n')}

      ${refs
        .map((refName) => `@ViewChild('${refName}') ${refName}: ElementRef`)
        .join('\n')}

      ${
        !component.hooks.onMount
          ? ''
          : `ngOnInit() {
              ${stripStateAndPropsRefs(component.hooks.onMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${
        !component.hooks.onUpdate?.length
          ? ''
          : `ngAfterContentChecked() {
              ${component.hooks.onUpdate.map((hook) =>
                stripStateAndPropsRefs(hook.code, {
                  replaceWith: 'this.',
                }),
              )}
            }`
      }

      ${
        !component.hooks.onUnMount
          ? ''
          : `ngOnDestroy() {
              ${stripStateAndPropsRefs(component.hooks.onUnMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${dataString}
    }
  `;

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };

const tryFormat = (str: string, parser: string) => {
  try {
    return format(str, {
      parser,
      plugins: [
        // To support running in browsers
        require('prettier/parser-typescript'),
        require('prettier/parser-postcss'),
        require('prettier/parser-html'),
        require('prettier/parser-babel'),
      ],
      htmlWhitespaceSensitivity: 'ignore',
    });
  } catch (err) {
    console.warn('Could not prettify', { string: str }, err);
  }
  return str;
};
