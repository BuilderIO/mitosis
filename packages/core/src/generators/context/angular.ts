import { stringifyContextValue } from '@/helpers/get-state-object-string';
import { MitosisContext } from '@/types/mitosis-context';
import { format } from 'prettier/standalone';

type ContextToAngularOptions = {
  format?: boolean;
  typescript?: boolean;
};

export const contextToAngular =
  (options: ContextToAngularOptions = { typescript: false }) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
import { Injectable } from '@angular/core';

@Injectable({
    providedIn: 'root'
})
export default class ${context.name}Context {
    ${stringifyContextValue(context.value)
      .replace(/^\{/, '')
      .replace(/\}$/, '')
      .replaceAll(',', ';\n')
      .replaceAll(':', ': any = ')}
    constructor() { }
}
  `;

    if (options.format !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
