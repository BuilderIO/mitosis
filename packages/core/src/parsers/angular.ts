import ts from 'typescript';
import { parseTemplate } from '@angular/compiler';
import {
  Node,
  Element,
  Template,
  Text,
  BoundText,
} from '@angular/compiler/src/render3/r3_ast';
import { ASTWithSource } from '@angular/compiler/src/expression_parser/ast';
import { createMitosisComponent } from '../helpers/create-mitosis-component';
import { createMitosisNode } from '../helpers/create-mitosis-node';
import { MitosisNode } from '../types/mitosis-node';
import { omit } from 'lodash';
import { babelTransformCode } from '../helpers/babel-transform';
import { types } from '@babel/core';
import { MitosisComponent } from '../types/mitosis-component';
import { capitalize } from '../helpers/capitalize';

const getTsAST = (code: string) => {
  return ts.createSourceFile('code.ts', code, ts.ScriptTarget.Latest, true);
};

interface AngularToMitosisOptions {}

const transformBinding = (
  binding: string,
  _options: AngularToMitosisOptions,
) => {
  return babelTransformCode(binding, {
    Identifier(path) {
      const name = path.node.name;

      if (
        (types.isObjectProperty(path.parent) &&
          path.parent.key === path.node) ||
        (types.isMemberExpression(path.parent) &&
          path.parent.property === path.node)
      ) {
        return;
      }

      if (
        !(name.startsWith('state.') || name === 'event' || name === '$event')
      ) {
        path.replaceWith(types.identifier(`state.${name}`));
      }
    },
  });
};

const isElement = (node: Node): node is Element =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).attributes);

const isTemplate = (node: Node): node is Template =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).templateAttrs);

const isText = (node: Node): node is Text =>
  typeof (node as any).value === 'string';

const isBoundText = (node: Node): node is BoundText =>
  typeof (node as any).value === 'object';

const angularTemplateNodeToMitosisNode = (
  node: Node,
  options: AngularToMitosisOptions,
): MitosisNode => {
  if (isTemplate(node)) {
    const ngIf = node.templateAttrs.find((item) => item.name === 'ngIf');
    if (ngIf) {
      return createMitosisNode({
        name: 'Show',
        bindings: {
          when: {
            code: transformBinding(
              (ngIf.value as ASTWithSource).source!,
              options,
            ),
          },
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
    const ngFor = node.templateAttrs.find((item) => item.name === 'ngFor');
    if (ngFor) {
      const value = (ngFor.value as ASTWithSource).source!;
      const split = value.split(/let\s|\sof\s/);
      const [_let, itemName, _of, expression] = split;
      return createMitosisNode({
        name: 'For',
        bindings: {
          each: { code: transformBinding(expression, options) },
        },
        properties: {
          _forName: itemName,
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
  }

  if (isElement(node)) {
    const properties: Record<string, string> = {};
    const bindings: Record<string, { code: string; arguments?: string[] }> = {};

    for (const input of node.inputs) {
      bindings[input.name] = {
        code: transformBinding((input.value as ASTWithSource).source!, options),
      };
    }
    for (const output of node.outputs) {
      bindings['on' + capitalize(output.name)] = {
        code: transformBinding(
          (output.handler as ASTWithSource)
            .source! // TODO: proper reference replace
            .replace(/\$event/g, 'event'),
          options,
        ),
      };
    }
    for (const attribute of node.attributes) {
      properties[attribute.name] = attribute.value;
    }

    return createMitosisNode({
      name: node.name,
      properties,
      bindings,
      children: node.children.map((node) =>
        angularTemplateNodeToMitosisNode(node, options),
      ),
    });
  }

  if (isText(node)) {
    return createMitosisNode({
      properties: {
        _text: node.value,
      },
    });
  }

  if (isBoundText(node)) {
    // TODO: handle the bindings
    return createMitosisNode({
      properties: {
        _text: (node.value as ASTWithSource).source!,
      },
    });
  }

  throw new Error(`Element node type {${node}} is not supported`);
};

const angularTemplateToMitosisNodes = (
  template: string,
  options: AngularToMitosisOptions,
) => {
  const ast = parseTemplate(template, '.');
  const blocks = ast.nodes.map((node) =>
    angularTemplateNodeToMitosisNode(node, options),
  );

  return blocks;
};

const parseTypescript = (code: string, options: AngularToMitosisOptions) => {
  const component = createMitosisComponent();

  const ast = getTsAST(code);
  for (const statement of ast.statements) {
    if (ts.isClassDeclaration(statement)) {
      if (statement.decorators) {
        for (const decorator of statement.decorators) {
          // TODO: proper reference tracing
          if (ts.isCallExpression(decorator.expression))
            if (
              ts.isIdentifier(decorator.expression.expression) &&
              decorator.expression.expression.text === 'Component'
            ) {
              const firstArg = decorator.expression.arguments[0];
              if (ts.isObjectLiteralExpression(firstArg)) {
                firstArg.properties.find((item) => {
                  if (ts.isPropertyAssignment(item)) {
                    if (
                      ts.isIdentifier(item.name) &&
                      item.name.text === 'template'
                    ) {
                      if (ts.isTemplateLiteral(item.initializer)) {
                        const template = item.initializer
                          .getText()
                          .trim()
                          .slice(1, -1);
                        component.children = angularTemplateToMitosisNodes(
                          template,
                          options,
                        );
                      }
                    }
                  }
                });
              }
            }
        }
      }
    }
  }

  return component;
};

export function angularToMitosisComponent(
  code: string,
  options: AngularToMitosisOptions = {},
): MitosisComponent {
  return parseTypescript(code, options);
}
