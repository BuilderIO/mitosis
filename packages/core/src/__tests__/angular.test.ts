import { componentToAngular } from '../generators/angular';
import { parseJsx } from '../parsers/jsx';

const onUpdate = require('./data/blocks/onUpdate.raw');
const onMount = require('./data/blocks/onMount.raw');
const onInitonMount = require('./data/blocks/onInit-onMount.raw');
const onInit = require('./data/blocks/onInit.raw');
const basicFor = require('./data/basic-for.raw');
const slot = require('./data/blocks/slot.raw');

describe('Angular', () => {
  test('onUpdate', () => {
    const component = parseJsx(onUpdate);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });

  test('onMount', () => {
    const component = parseJsx(onMount);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });

  test('onInit and onMount', () => {
    const component = parseJsx(onInitonMount);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });

  test('onInit', () => {
    const component = parseJsx(onInit);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });

  test('BasicFor', () => {
    const component = parseJsx(basicFor);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });

  test('ng-content', () => {
    const component = parseJsx(slot);
    const output = componentToAngular()({ component });
    expect(output).toMatchSnapshot();
  });
});
