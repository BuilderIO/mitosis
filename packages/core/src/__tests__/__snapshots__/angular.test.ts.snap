// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Angular BasicFor 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"my-basic-for-component\\",
  template: \`
    <div>
      <ng-container *ngFor=\\"let person of names\\">
        <div>
          <input
            [value]=\\"name\\"
            (input)=\\"
         name = $event.target.value + ' and ' + person;
       \\"
          />

          Hello {{person}} ! I can run in Qwik, Web Component, React, Vue,
          Solid, or Liquid!
        </div>
      </ng-container>
    </div>
  \`,
})
export default class MyBasicForComponent {
  ngOnInit() {
    console.log(\\"onMount code\\");
  }

  name = \\"PatrickJS\\";
  names = [\\"Steve\\", \\"PatrickJS\\"];
}
"
`;

exports[`Angular Slot 1`] = `
"import { Component } from \\"@angular/core\\";

import ContentSlotCode from \\"./content-slot.raw\\";

@Component({
  selector: \\"slot-code\\",
  template: \`
    <div>
      <ContentSlotCode><div testing>Hello</div></ContentSlotCode>
    </div>
  \`,
})
export default class SlotCode {}
"
`;

exports[`Angular ng-content and Slot 1`] = `
"import { Component, Input } from \\"@angular/core\\";

@Component({
  selector: \\"content-slot-code\\",
  template: \`
    <div>
      <ng-content select=\\"[testing]\\"></ng-content>

      <div>
        <hr />
      </div>

      <div><ng-content></ng-content></div>
    </div>
  \`,
})
export default class ContentSlotCode {
  @Input() children: any;
}
"
`;

exports[`Angular onInit 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"on-init\\",
  template: \`
    <div></div>
  \`,
})
export default class OnInit {
  ngOnInit() {
    console.log(\\"Runs once every update/rerender\\");
  }
}
"
`;

exports[`Angular onInit and onMount 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"on-init\\",
  template: \`
    <div></div>
  \`,
})
export default class OnInit {
  ngOnInit() {
    console.log(\\"onInit\\");

    console.log(\\"onMount\\");
  }
}
"
`;

exports[`Angular onMount 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"comp\\",
  template: \`
    <div></div>
  \`,
})
export default class Comp {
  ngOnInit() {
    console.log(\\"Runs on mount\\");
  }

  ngOnDestroy() {
    console.log(\\"Runs on unMount\\");
  }
}
"
`;

exports[`Angular onUpdate 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"on-update\\",
  template: \`
    <div></div>
  \`,
})
export default class OnUpdate {
  ngAfterContentChecked() {
    console.log(\\"Runs on every update/rerender\\");
  }
}
"
`;
